ParmEd i/o

lilyminium opened the pull request:
https://github.com/MDAnalysis/mdanalysis/pull/2381#issuecomment-550104207
> I'd rather look into the question how MDAnalysis can call ParmEd to write a topology.

Library interoperability is convenient. There are a few issues:

- ParmEd doesn't try to guess elements if they aren't read. The current ParmEdParser does. This means, e.g., the PSF test file can't roundtrip. This could be an advantage for when you want a Parmed structure with the right elements

- ParmEd's guessing isn't great. It thinks that NA+ has an atomic number of 7, for example. 

- MD engines and ParmEd support multiple bonds etc., while MDAnalysis doesn't. Currently ParmEdParser turns these into a tuple of bond(etc.)types

**Example**
```python
In [1]: import MDAnalysis as mda

In [2]: import parmed as pmd

In [3]: from MDAnalysis.tests.datafiles import *

In [4]: pgro = pmd.load_file(GRO)

In [5]: mgro = mda.Universe(pgro)

In [6]: ogro = mgro.atoms.write(file_format='PARMED')
```

Changes made in this Pull Request:
 - Added ParmEdParser
 - Added ParmEdReader
 - Added ParmEdWriter
 - Added UreyBradleys + CMaps


PR Checklist
------------
 - [x] Tests? (some)
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

Still missing tests for:
- [x] Writing a subset of atoms
- [x] Writing an MDAnalysis Universe read from a non-Parmed source
- [x] Anything with Urey-bradley angles?

richardjgowers commented on package/MDAnalysis/coordinates/ParmEd.py:
```diff
@@ -0,0 +1,264 @@
+    'id': 'number'
+}
+
+SYMB2Z = {'H': 1,                                                   'He': 2, 
```
this should probably go somewhere like `topology.tables` or something

richardjgowers commented on .appveyor.yml:
```diff
@@ -9,8 +9,8 @@ cache:
 
 environment:
     global:
-        CONDA_CHANNELS: conda-forge
```
I'd rather move the parmed package to conda-forge

orbeckst replied:
I agree.

Can both be used together, with conda-forge higher priority? IIRC, conda's package resolving process gets slower and slower with more channels so we really try to minimize time there.

orbeckst replied:
Btw, omnia requires conda-forge https://github.com/ParmEd/ParmEd/blob/7e411fd03c7db6977e450c2461e065004adab471/devtools/ci/travis/install.sh#L25-L26

orbeckst replied:
The full dependency list from https://github.com/ParmEd/ParmEd/blob/7e411fd03c7db6977e450c2461e065004adab471/devtools/ci/travis/install.sh#L25-L39
```bash
    if [ -z "$MINIMAL_PACKAGES" ]; then
        # Install all prerequisites
        conda create -y -n myenv python=$PYTHON_VERSION \
            numpy scipy pandas nose openmm coverage nose-timer \
            netCDF4
        conda update -y -n myenv --all
        conda install -y -n myenv rdkit==2018.09.1
        conda install -y -n myenv boost==1.69.0
        conda install -y -n myenv nglview
        conda install -y -n myenv ambertools=18 -c ambermd
        conda install -y -n myenv networkx
        conda install -y -n myenv lxml
```
Note that it contains ambertools from the ambermd channel, rdkit, and boost as some heavy dependencies.

The minimal list https://github.com/ParmEd/ParmEd/blob/7e411fd03c7db6977e450c2461e065004adab471/devtools/ci/travis/install.sh#L41-L43 is essentially numpy:
```bash
        conda create -y -n myenv python=$PYTHON_VERSION numpy nose coverage \
            nose-timer
```

I don't really want any MDA user have to install these massive dependencies. But I also don't think one should create a crippled ParmEd conda-forge package (which does not pull all its dependencies) as this is a bad user experience for anyone who installs ParmEd for its own sake.

Should we make the ParmEdReader/Writer fail gracefully if no parmed package is installed? This would, however, go against our [importing rules](https://github.com/MDAnalysis/mdanalysis/wiki/Style-Guide#importing-modules ) for [modules in the core](https://github.com/MDAnalysis/mdanalysis/wiki/Style-Guide#list-of-core-module-dependencies).

orbeckst replied:
Regarding "guidelines" vs "rules": https://www.youtube.com/watch?v=WJVBvvS57j0 – should our rules be more something that we would call "guidelines"?

mattwthompson replied:
Yes and (painfully) yes

`omnia` is a pretty well-maintained channel, although fairly specialized to the projects of the Pande, Chodera, Mobley, etc. labs. I'm not sure who is responsible for it but it also happens to be packaged on [`conda-forge`](https://anaconda.org/conda-forge/parmed). Compared to other dependencies, ParmEd is updated quite infrequently, so the risk of upstream complications should be fairly low (of course, there's also the downside of it possibly being slow to update to newer Python versions; not sure how that fits into your plans).

hainm replied:
@orbeckst parmed only requires `numpy`. And it's pretty easy to install via pip (or conda if exclude the requirements).

orbeckst replied:
That's good to know.

Thanks for comments, @mattihappy and @hainm !

richardjgowers replied:
Looks like the conda-forge recipe only does linux for now, I raised an issue

richardjgowers commented on testsuite/MDAnalysisTests/coordinates/test_parmed.py:
```diff
@@ -0,0 +1,153 @@
+    
+
+class TestParmEdReaderPDB(BaseTestParmEdReader):
+    __test__ = True
```
shouldn't need `__test__`

orbeckst commented on package/MDAnalysis/coordinates/ParmEd.py:
```diff
@@ -0,0 +1,264 @@
+    units = {'time': None, 'length': 'Angstrom'}
+
+    def __init__(self, n_atoms=None, convert_units=None, **kwargs):
+        """Set up a ParmEdWriter with a precision of 3 decimal places.
```
Why "precision of 3 decimal places" ?

lilyminium replied:
I have no idea. Leftover from something I copied? No precision is ever considered, anyway. Removed.

orbeckst commented on package/MDAnalysis/coordinates/ParmEd.py:
```diff
@@ -0,0 +1,264 @@
+================================================================
+
+Reads coordinates data from a ParmEd Structure.
+
```
Minimal usage example would be really good.

orbeckst commented on package/MDAnalysis/coordinates/ParmEd.py:
```diff
@@ -0,0 +1,264 @@
+
+
+class ParmEdWriter(base.WriterBase):
+    """Convert MDAnalysis AtomGroup or Universe to ParmEd Structure
```
That's a very interesting interoperability idea!

orbeckst commented on package/MDAnalysis/core/topologyattrs.py:
```diff
@@ -1286,6 +1286,82 @@ class AltLocs(AtomAttr):
     def _gen_initial_values(na, nr, ns):
         return np.array(['' for _ in range(na)], dtype=object)
 
+class GBScreens(AtomAttr):
```
Being able to add all these attributes shows off the flexibility of the topology system!

orbeckst commented on package/MDAnalysis/topology/ParmEdParser.py:
```diff
@@ -0,0 +1,317 @@
+Example
+-------
+
+>>> import parmed as pmd
```
This is a writing example. Can you also add a reading example?

orbeckst commented on package/MDAnalysis/topology/ParmEdParser.py:
```diff
@@ -0,0 +1,317 @@
+
+.. _ParmEd: https://parmed.github.io/ParmEd/html/index.html
+
+Example
```
I *think* such a heading will not render nicely with numpy docs/napoleon. We normally rewrite it to not be the word "Example" or "Examples", such as "Example use" or something more meaningful.

(I thought we had a note regarding this problem at https://github.com/MDAnalysis/mdanalysis/wiki/WritingDocumentation#numpy-style-gotchas but we don't so I am not 100% sure if this is still a problem.)

lilyminium replied:
I think normally the heading used in MDAnalysis is just Example/s? It displays alright for [Universe](https://www.mdanalysis.org/mdanalysis/documentation_pages/core/universe.html#MDAnalysis.core.universe.Universe.add_TopologyAttr).

orbeckst commented on package/MDAnalysis/topology/ParmEdParser.py:
```diff
@@ -0,0 +1,317 @@
+
+# >>> import mendeleev as mv
+# >>> {i:mv.element(i).symbol for i in range(1, 100)}
+Z2SYMB = {1: 'H',                                                    2: 'He', 
```
Can't you generate this table from `coordinates.ParmEd.SYMB2Z` ? 

Furthermore, perhaps just putting the periodic table into `lib.tables` might be a better place anyway?

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_parmed.py:
```diff
@@ -0,0 +1,153 @@
+    __test__ = True
+    ref_filename = GRO
+
+    universe = mda.Universe(pmd.load_file(GRO))
```
Add this as an example to the ParmEdReader and Parser!

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_parmed.py:
```diff
@@ -0,0 +1,153 @@
+            for attr in ('mass', 'charge', 'occupancy', 'rmin', 'epsilon'):
+                ra = getattr(r, attr)
+                oa = getattr(o, attr)
+                assert abs(ra-oa) < self.prec
```
Don't roll your own – just use `assert_almost_equal`

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_parmed.py:
```diff
@@ -0,0 +1,153 @@
+    def test_coordinates(self):
+        up = self.universe.atoms.positions
+        rp = self.ref.atoms.positions
+        assert_array_almost_equal(up, rp, decimal=3)
```
We stopped using `assert_array_almost_equal` and replaced it with `assert_almost_equal` — although I forget the reason...

orbeckst commented:
I _think_ converters are a good idea, especially on the writing side. Nevertheless, one could also argue that anything that takes in a data structure/class with a given API should be called a converter instead of a Reader. Once this is in, we could classify MMTF and MemoryReader as a converters as they all take a data structure as input as opposed to a filename.

richardjgowers commented on package/MDAnalysis/core/groups.py:
```diff
@@ -3032,6 +3032,82 @@ def improper(self):
+        
+        """
+        try:
+            writer = _CONVERTERS[format]
```
Rather than exposing (and pinning down) how we're doing the magic around `_CONVERTERS`, it's better to have a `get_converter_for` method like writers have, where later we can put more/different magic if things change.

lilyminium replied:
Thanks, moved.

orbeckst commented on package/MDAnalysis/core/groups.py:
```diff
@@ -3032,6 +3032,80 @@ def improper(self):
+            import MDAnalysis as mda
+            from MDAnalysis.tests.datafiles import GRO
+            u = mda.Universe(GRO)
+            parmed_structure = u.atoms.convert_to('PARMED')
```
Add sentence stating that `parmed_structure` is now a [ParmEd.structure.Structure](https://parmed.github.io/ParmEd/html/structobj/parmed.structure.Structure.html) instance.

Link to the ParmEd doc.

Add an entry for https://parmed.github.io/ParmEd/html/ to the intersphinx links
https://github.com/MDAnalysis/mdanalysis/blob/972a8b1893b5e353b9d796b15d6fcdd022a93f00/package/doc/sphinx/source/conf.py#L329-L336

orbeckst commented on package/MDAnalysis/topology/ParmEdParser.py:
```diff
@@ -0,0 +1,303 @@
+    import MDAnalysis as mda
+    from MDAnalysis.tests.datafiles import PSF
+    universe = mda.Universe(pmd.load_file(PSF))
+
```
See comments above – say something about the resulting object (ParmEd.structure.Structure).

orbeckst commented on package/MDAnalysis/topology/__init__.py:
```diff
@@ -312,3 +312,4 @@
 from . import GSDParser
 from . import MinimalParser
 from . import ITPParser
+from . import ParmEdParser
```
newline

orbeckst commented on package/MDAnalysis/topology/tables.py:
```diff
@@ -377,3 +377,20 @@ def kv2dict(s, convertor=str):
+          75: 'Re', 76: 'Os', 77: 'Ir', 78: 'Pt', 79: 'Au', 80: 'Hg', 81: 'Tl', 82: 'Pb', 
+          83: 'Bi', 84: 'Po', 85: 'At', 86: 'Rn', 87: 'Fr', 88: 'Ra', 89: 'Ac', 90: 'Th', 
+          91: 'Pa', 92: 'U',  93: 'Np', 94: 'Pu', 95: 'Am', 96: 'Cm', 97: 'Bk', 98: 'Cf', 
+          99: 'Es'}
```
Why stop at 99?  ;-)

orbeckst commented on package/MDAnalysis/topology/tables.py:
```diff
@@ -377,3 +377,20 @@ def kv2dict(s, convertor=str):
+          91: 'Pa', 92: 'U',  93: 'Np', 94: 'Pu', 95: 'Am', 96: 'Cm', 97: 'Bk', 98: 'Cf', 
+          99: 'Es'}
+
+SYMB2Z = {v:k for k, v in Z2SYMB.items()}
```
newline.

richardjgowers commented:
@orbeckst yeah I'm not sure where the line on converters is, but maybe a nice definiton is that a converter is moving from one package's namespace/object to another.

richardjgowers commented on .appveyor.yml:
```diff
@@ -9,8 +9,8 @@ cache:
     global:
-        CONDA_CHANNELS: conda-forge
-        CONDA_DEPENDENCIES: pip setuptools wheel cython mock six biopython networkx joblib matplotlib scipy vs2015_runtime pytest mmtf-python GridDataFormats hypothesis pytest-cov codecov
+        CONDA_CHANNELS: conda-forge omnia
```
parmed is in conda-forge

orbeckst replied:
Then drop omnia, let us know what dependencies it pulls in, and then I am happy.

lilyminium commented:
Thanks @orbeckst. Added a couple usage examples and took out the `omnia` channel.

> Maybe we don't need to reorganize their code but their docs should be prominent.

I added a [brief page on converters](https://lilyminium.github.io/mdanalysis/documentation_pages/converters.html) -- is that what you meant?

orbeckst commented:
> I added a brief page on converters -- is that what you meant?

Yes, that's a good starting point!

orbeckst commented:
I approved it because you addressed my concerns – thank you.

Obviously, tests should be passing, too. I am not sure why many of the last tests failed but I restarted them. If they fail again, have a look...
